File: k8s/meilisearch/index-examples.py
#!/usr/bin/env python3
"""
Indexes Aidbox examples into Meilisearch from GitHub Actions artifact.

This script:
1. Fetches the latest examples-metadata artifact from GitHub Actions
2. Downloads and extracts examples-metadata.json from the ZIP
3. Configures the Meilisearch "examples" index
4. Indexes all examples documents

Environment variables required:
- MEILISEARCH_HOST_URL: Meilisearch server URL
- MEILISEARCH_API_KEY: Meilisearch API key
- GITHUB_TOKEN: GitHub Personal Access Token for API access
"""

import os
import sys
import json
import time
from io import BytesIO
from zipfile import ZipFile
from urllib.request import Request, urlopen, build_opener, HTTPRedirectHandler
from urllib.error import HTTPError, URLError


class NoRedirectHandler(HTTPRedirectHandler):
    """Custom handler that doesn't follow redirects"""
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def log(message, **kwargs):
    """Log message with optional key-value pairs"""
    if kwargs:
        extras = " ".join(f"{k}={v}" for k, v in kwargs.items())
        print(f"{message} {extras}", flush=True)
    else:
        print(message, flush=True)


def validate_env():
    """Validate required environment variables"""
    required = ["MEILISEARCH_HOST_URL", "MEILISEARCH_API_KEY", "GITHUB_TOKEN"]
    missing = [var for var in required if not os.getenv(var)]

    if missing:
        log("ERROR: Missing required environment variables:", vars=", ".join(missing))
        sys.exit(1)

    log("Environment variables validated")


def fetch_latest_artifact(github_token):
    """
    Fetch list of artifacts and return the latest examples-metadata artifact.

    Returns:
        dict: Artifact metadata with 'id' and 'created_at' fields, or None if not found
    """
    url = "https://api.github.com/repos/Aidbox/examples/actions/artifacts"
    headers = {
        "Authorization": f"Bearer {github_token}",
        "Accept": "application/vnd.github.v3+json"
    }

    try:
        log("Fetching artifacts list from GitHub...")
        req = Request(url, headers=headers)
        with urlopen(req, timeout=10) as response:
            data = json.loads(response.read().decode())

        artifacts = data.get("artifacts", [])
        log(f"Found {len(artifacts)} total artifacts")

        # Filter for examples-metadata artifacts that are not expired
        examples_artifacts = [
            a for a in artifacts
            if a.get("name") == "examples-metadata" and not a.get("expired", False)
        ]

        if not examples_artifacts:
            log("WARNING: No valid examples-metadata artifacts found")
            return None

        # Sort by created_at and get the latest
        latest = sorted(examples_artifacts, key=lambda x: x.get("created_at", ""))[-1]
        log(f"Latest artifact found:",
            artifact_id=latest["id"],
            created_at=latest.get("created_at"))

        return latest

    except HTTPError as e:
        log(f"ERROR: GitHub API HTTP error: {e.code} {e.reason}")
        sys.exit(1)
    except URLError as e:
        log(f"ERROR: Network error fetching artifacts: {e.reason}")
        sys.exit(1)
    except Exception as e:
        log(f"ERROR: Failed to fetch artifacts: {e}")
        sys.exit(1)


def download_artifact(github_token, artifact_id):
    """
    Download artifact ZIP from GitHub (handles 302 redirect to Azure).

    Args:
        github_token: GitHub PAT
        artifact_id: Artifact ID to download

    Returns:
        bytes: ZIP file contents
    """
    url = f"https://api.github.com/repos/Aidbox/examples/actions/artifacts/{artifact_id}/zip"
    headers = {
        "Authorization": f"Bearer {github_token}",
        "Accept": "application/vnd.github.v3+json"
    }

    try:
        log(f"Downloading artifact {artifact_id}...")

        # First request - GitHub will return 302 redirect
        # Use custom opener that doesn't follow redirects
        opener = build_opener(NoRedirectHandler())
        req = Request(url, headers=headers)

        try:
            response = opener.open(req, timeout=30)
            zip_bytes = response.read()
            log(f"Artifact downloaded directly:", size=len(zip_bytes))
            return zip_bytes
        except HTTPError as e:
            if e.code == 302:
                # Get redirect location
                redirect_url = e.headers.get("Location")
                if not redirect_url:
                    log("ERROR: Got 302 but no Location header")
                    sys.exit(1)

                log("Following redirect to Azure storage...")

                # Follow redirect WITHOUT Authorization header (Azure doesn't need it)
                redirect_req = Request(redirect_url)
                with urlopen(redirect_req, timeout=30) as redirect_response:
                    zip_bytes = redirect_response.read()
                    log(f"Artifact downloaded from redirect:", size=len(zip_bytes))
                    return zip_bytes
            else:
                raise

    except HTTPError as e:
        log(f"ERROR: HTTP error downloading artifact: {e.code} {e.reason}")
        error_body = e.read().decode() if e.fp else ""
        if error_body:
            log(f"Error details: {error_body}")
        sys.exit(1)
    except URLError as e:
        log(f"ERROR: Network error downloading artifact: {e.reason}")
        sys.exit(1)
    except Exception as e:
        log(f"ERROR: Failed to download artifact: {e}")
        sys.exit(1)


def extract_metadata(zip_bytes):
    """
    Extract examples-metadata.json from ZIP bytes.

    Args:
        zip_bytes: ZIP file as bytes

    Returns:
        dict: Parsed JSON data
    """
    try:
        log("Extracting examples-metadata.json from ZIP...")

        with ZipFile(BytesIO(zip_bytes)) as zip_file:
            # List files in ZIP
            file_list = zip_file.namelist()
            log(f"ZIP contains {len(file_list)} files")

            if "examples-metadata.json" not in file_list:
                log("ERROR: examples-metadata.json not found in ZIP")
                log("Files in ZIP:", files=", ".join(file_list))
                sys.exit(1)

            # Extract and parse JSON
            with zip_file.open("examples-metadata.json") as json_file:
                data = json.load(json_file)

            examples_count = len(data.get("examples", []))
            log(f"Extracted metadata:", examples_count=examples_count)

            return data

    except json.JSONDecodeError as e:
        log(f"ERROR: Invalid JSON in metadata file: {e}")
        sys.exit(1)
    except Exception as e:
        log(f"ERROR: Failed to extract metadata: {e}")
        sys.exit(1)


def transform_for_meilisearch(examples_data):
    """
    Transform examples data for Meilisearch indexing.

    Args:
        examples_data: Dict with 'examples', 'features_list', 'languages_list'

    Returns:
        list: Documents ready for indexing
    """
    log("Transforming data for Meilisearch...")

    examples = examples_data.get("examples", [])

    if not examples:
        log("WARNING: No examples found in metadata")
        return []

    # Keep all fields as-is
    # Expected fields: id, title, description, category, github_url, features, languages
    documents = []

    for example in examples:
        # Validate required fields
        if not example.get("id"):
            log("WARNING: Skipping example without id:", example=example)
            continue

        if not example.get("title"):
            log("WARNING: Skipping example without title:", id=example.get("id"))
            continue

        # Fix ID: Meilisearch doesn't allow slashes in document IDs
        # Replace / with _ (underscore)
        example_copy = example.copy()
        example_copy["id"] = example_copy["id"].replace("/", "_")

        documents.append(example_copy)

    log(f"Prepared {len(documents)} documents for indexing")
    return documents


def make_meilisearch_request(url, method, headers, data=None, retry_count=3):
    """
    Make HTTP request to Meilisearch with retry logic.

    Args:
        url: Request URL
        method: HTTP method (GET, POST, PUT, etc.)
        headers: Request headers dict
        data: Optional request body (will be JSON encoded)
        retry_count: Number of retries on failure

    Returns:
        dict: Response data (parsed JSON) or None
    """
    for attempt in range(retry_count):
        try:
            req_data = json.dumps(data).encode() if data else None
            req = Request(url, data=req_data, headers=headers, method=method)

            with urlopen(req, timeout=30) as response:
                if response.status in [200, 201, 202]:
                    response_data = response.read().decode()
                    return json.loads(response_data) if response_data else {}
                else:
                    log(f"WARNING: Unexpected status {response.status}")
                    return None

        except HTTPError as e:
            if e.code == 404 and method == "GET":
                # Index doesn't exist yet, this is OK
                return None
            log(f"Attempt {attempt + 1}/{retry_count} failed:",
                error=f"{e.code} {e.reason}")
            if attempt < retry_count - 1:
                time.sleep(2 ** attempt)  # Exponential backoff
            else:
                raise
        except URLError as e:
            log(f"Attempt {attempt + 1}/{retry_count} failed:", error=str(e.reason))
            if attempt < retry_count - 1:
                time.sleep(2 ** attempt)
            else:
                raise


def configure_index(meilisearch_url, api_key):
    """
    Create and configure the examples index in Meilisearch.

    Args:
        meilisearch_url: Meilisearch server URL
        api_key: Meilisearch API key
    """
    log("Configuring Meilisearch index...")

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    # Create index if it doesn't exist
    index_url = f"{meilisearch_url}/indexes"
    index_data = {
        "uid": "examples",
        "primaryKey": "id"
    }

    try:
        # Check if index exists
        check_url = f"{meilisearch_url}/indexes/examples"
        existing = make_meilisearch_request(check_url, "GET", headers)

        if not existing:
            log("Creating new index...")
            make_meilisearch_request(index_url, "POST", headers, index_data)
            log("Index created")
        else:
            log("Index already exists")

        # Configure index settings
        settings_url = f"{meilisearch_url}/indexes/examples/settings"
        settings_data = {
            "searchableAttributes": [
                "title",
                "description",
                "category",
                "features",
                "languages"
            ],
            "filterableAttributes": [
                "category",
                "features",
                "languages"
            ],
            "displayedAttributes": [
                "id",
                "title",
                "description",
                "category",
                "github_url",
                "features",
                "languages"
            ],
            "rankingRules": [
                "words",
                "typo",
                "proximity",
                "attribute",
                "sort",
                "exactness"
            ]
        }

        log("Updating index settings...")
        make_meilisearch_request(settings_url, "PATCH", headers, settings_data)
        log("Index settings updated")

    except Exception as e:
        log(f"ERROR: Failed to configure index: {e}")
        sys.exit(1)


def index_documents(meilisearch_url, api_key, documents):
    """
    Index documents into Meilisearch examples index.

    Args:
        meilisearch_url: Meilisearch server URL
        api_key: Meilisearch API key
        documents: List of documents to index
    """
    if not documents:
        log("No documents to index")
        return

    log(f"Indexing {len(documents)} documents...")

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    try:
        url = f"{meilisearch_url}/indexes/examples/documents"
        response = make_meilisearch_request(url, "POST", headers, documents)

        if response:
            task_uid = response.get("taskUid")
            log("Documents indexed successfully:", task_uid=task_uid)
        else:
            log("WARNING: Indexing may have failed (no response)")

    except Exception as e:
        log(f"ERROR: Failed to index documents: {e}")
        sys.exit(1)


def main():
    """Main execution flow"""
    log("=" * 60)
    log("Starting Aidbox Examples Indexer")
    log("=" * 60)

    # Step 1: Validate environment
    validate_env()

    meilisearch_url = os.getenv("MEILISEARCH_HOST_URL").rstrip("/")
    api_key = os.getenv("MEILISEARCH_API_KEY")
    github_token = os.getenv("GITHUB_TOKEN")

    # Step 2: Fetch latest artifact
    artifact = fetch_latest_artifact(github_token)
    if not artifact:
        log("No artifact to process, exiting")
        sys.exit(0)

    # Step 3: Download artifact ZIP
    zip_bytes = download_artifact(github_token, artifact["id"])

    # Step 4: Extract metadata
    examples_data = extract_metadata(zip_bytes)

    # Step 5: Transform data
    documents = transform_for_meilisearch(examples_data)

    # Step 6: Configure index
    configure_index(meilisearch_url, api_key)

    # Step 7: Index documents
    index_documents(meilisearch_url, api_key, documents)

    log("=" * 60)
    log("Indexing completed successfully")
    log("=" * 60)


if __name__ == "__main__":
    main()

File: k8s/meilisearch/index-examples.ts
#!/usr/bin/env bun
/**
 * Indexes Aidbox examples into Meilisearch from GitHub Actions artifact.
 *
 * This script:
 * 1. Fetches the latest examples-metadata artifact from GitHub Actions
 * 2. Downloads and extracts examples-metadata.json from the ZIP
 * 3. Creates temp index and indexes all examples documents
 * 4. Swaps temp index with main index (zero downtime)
 * 5. Deletes old temp index
 *
 * Environment variables required:
 * - MEILISEARCH_HOST_URL: Meilisearch server URL
 * - MEILISEARCH_API_KEY: Meilisearch API key
 * - GITHUB_TOKEN: GitHub Personal Access Token for API access
 */

import { MeiliSearch } from "meilisearch";

interface ExampleDocument {
  id: string;
  title: string;
  description: string;
  category: string;
  github_url: string;
  readme_url: string;
  features: string[];
  languages: string[];
}

interface ExamplesMetadata {
  examples: ExampleDocument[];
  features_list: string[];
  languages_list: string[];
}

interface GitHubArtifact {
  id: number;
  name: string;
  created_at: string;
  expired: boolean;
}

function log(message: string, data?: Record<string, any>) {
  if (data) {
    const extras = Object.entries(data)
      .map(([k, v]) => `${k}=${v}`)
      .join(" ");
    console.log(`${message} ${extras}`);
  } else {
    console.log(message);
  }
}

function validateEnv() {
  const required = ["MEILISEARCH_HOST_URL", "MEILISEARCH_API_KEY", "GITHUB_TOKEN"];
  const missing = required.filter((v) => !process.env[v]);

  if (missing.length > 0) {
    log("ERROR: Missing required environment variables:", { vars: missing.join(", ") });
    process.exit(1);
  }

  log("Environment variables validated");
}

async function fetchLatestArtifact(githubToken: string): Promise<GitHubArtifact | null> {
  const url = "https://api.github.com/repos/Aidbox/examples/actions/artifacts";
  const headers = {
    Authorization: `Bearer ${githubToken}`,
    Accept: "application/vnd.github.v3+json",
  };

  try {
    log("Fetching artifacts list from GitHub...");
    const response = await fetch(url, { headers });

    if (!response.ok) {
      log(`ERROR: GitHub API HTTP error: ${response.status} ${response.statusText}`);
      process.exit(1);
    }

    const data = await response.json();
    const artifacts = data.artifacts || [];
    log(`Found ${artifacts.length} total artifacts`);

    // Filter for examples-metadata artifacts that are not expired
    const examplesArtifacts = artifacts.filter(
      (a: GitHubArtifact) => a.name === "examples-metadata" && !a.expired
    );

    if (examplesArtifacts.length === 0) {
      log("WARNING: No valid examples-metadata artifacts found");
      return null;
    }

    // Sort by created_at and get the latest
    const latest = examplesArtifacts.sort(
      (a: GitHubArtifact, b: GitHubArtifact) => a.created_at.localeCompare(b.created_at)
    )[examplesArtifacts.length - 1];

    log("Latest artifact found:", {
      artifact_id: latest.id,
      created_at: latest.created_at,
    });

    return latest;
  } catch (error) {
    log(`ERROR: Failed to fetch artifacts: ${error}`);
    process.exit(1);
  }
}

async function downloadArtifact(githubToken: string, artifactId: number): Promise<ArrayBuffer> {
  const url = `https://api.github.com/repos/Aidbox/examples/actions/artifacts/${artifactId}/zip`;
  const headers = {
    Authorization: `Bearer ${githubToken}`,
    Accept: "application/vnd.github.v3+json",
  };

  try {
    log(`Downloading artifact ${artifactId}...`);

    // First request - GitHub will return 302 redirect
    const response = await fetch(url, {
      headers,
      redirect: "manual",
    });

    if (response.status === 302) {
      const redirectUrl = response.headers.get("Location");
      if (!redirectUrl) {
        log("ERROR: Got 302 but no Location header");
        process.exit(1);
      }

      log("Following redirect to Azure storage...");

      // Follow redirect WITHOUT Authorization header (Azure doesn't need it)
      const redirectResponse = await fetch(redirectUrl);
      if (!redirectResponse.ok) {
        log(`ERROR: Failed to download from redirect: ${redirectResponse.status}`);
        process.exit(1);
      }

      const zipBytes = await redirectResponse.arrayBuffer();
      log(`Artifact downloaded from redirect:`, { size: zipBytes.byteLength });
      return zipBytes;
    }

    if (!response.ok) {
      log(`ERROR: Failed to download artifact: ${response.status}`);
      process.exit(1);
    }

    const zipBytes = await response.arrayBuffer();
    log(`Artifact downloaded directly:`, { size: zipBytes.byteLength });
    return zipBytes;
  } catch (error) {
    log(`ERROR: Failed to download artifact: ${error}`);
    process.exit(1);
  }
}

async function extractMetadata(zipBytes: ArrayBuffer): Promise<ExamplesMetadata> {
  try {
    log("Extracting examples-metadata.json from ZIP...");

    // Use Bun's built-in unzipping
    const decoder = new TextDecoder();
    const zipBlob = new Blob([zipBytes]);

    // Bun doesn't have built-in ZIP, so we'll use a workaround
    // Write to temp file and use unzipping
    const tempFile = `/tmp/examples-${Date.now()}.zip`;
    await Bun.write(tempFile, zipBlob);

    // Use shell unzip to extract
    const proc = Bun.spawn(["unzip", "-p", tempFile, "examples-metadata.json"]);
    const output = await new Response(proc.stdout).text();

    // Clean up temp file
    await Bun.spawn(["rm", tempFile]).exited;

    const data = JSON.parse(output) as ExamplesMetadata;
    log(`Extracted metadata:`, { examples_count: data.examples.length });

    return data;
  } catch (error) {
    log(`ERROR: Failed to extract metadata: ${error}`);
    process.exit(1);
  }
}

function transformForMeilisearch(examplesData: ExamplesMetadata): ExampleDocument[] {
  log("Transforming data for Meilisearch...");

  const examples = examplesData.examples || [];

  if (examples.length === 0) {
    log("WARNING: No examples found in metadata");
    return [];
  }

  const documents: ExampleDocument[] = [];

  for (const example of examples) {
    // Validate required fields
    if (!example.id) {
      log("WARNING: Skipping example without id:", { example });
      continue;
    }

    if (!example.title) {
      log("WARNING: Skipping example without title:", { id: example.id });
      continue;
    }

    // Meilisearch doesn't allow slashes in document IDs
    const fixedExample = {
      ...example,
      id: example.id.replace(/\//g, "_"),
    };

    documents.push(fixedExample);
  }

  log(`Prepared ${documents.length} documents for indexing`);
  return documents;
}

async function reindexExamples(
  client: MeiliSearch,
  indexName: string,
  documents: ExampleDocument[]
) {
  const tempIndexName = `${indexName}_temp`;

  log(`\nStarting reindex for: ${indexName}`);
  log(`Total documents: ${documents.length}`);

  // 1. Delete temp index if exists from previous failed run
  try {
    await client.deleteIndex(tempIndexName);
    log(`Deleted index: ${tempIndexName}`);
  } catch (e) {
    // Index doesn't exist, that's fine
  }

  // 2. Create temp index with settings
  log(`Creating index: ${tempIndexName}`);
  const createTask = await client.createIndex(tempIndexName, { primaryKey: "id" });
  await client.waitForTask(createTask.taskUid);

  // 3. Apply settings to temp index
  const settings = {
    searchableAttributes: ["title", "description", "category", "features", "languages"],
    filterableAttributes: ["category", "features", "languages"],
    displayedAttributes: [
      "id",
      "title",
      "description",
      "category",
      "github_url",
      "readme_url",
      "features",
      "languages",
    ],
    rankingRules: ["words", "typo", "proximity", "attribute", "sort", "exactness"],
  };

  log(`Applying settings to index: ${tempIndexName}`);
  const settingsTask = await client.index(tempIndexName).updateSettings(settings);
  await client.waitForTask(settingsTask.taskUid);

  // 4. Add documents to temp index in batches
  const batchSize = 100;
  for (let i = 0; i < documents.length; i += batchSize) {
    const batch = documents.slice(i, i + batchSize);
    log(`Indexing ${i + 1}-${Math.min(i + batchSize, documents.length)}/${documents.length}`);

    const addTask = await client.index(tempIndexName).addDocuments(batch);
    await client.waitForTask(addTask.taskUid);
  }

  // 5. Check if main index exists
  let mainExists = false;
  try {
    await client.getIndex(indexName);
    mainExists = true;
  } catch (e) {
    mainExists = false;
  }

  if (!mainExists) {
    // Create empty main index first
    log(`Creating main index: ${indexName}`);
    const mainCreateTask = await client.createIndex(indexName, { primaryKey: "id" });
    await client.waitForTask(mainCreateTask.taskUid);
  }

  // 6. Swap indexes (atomic operation)
  log(`Swapping indexes: ${indexName} <-> ${tempIndexName}`);
  const swapTask = await client.swapIndexes([{ indexes: [indexName, tempIndexName] }]);
  await client.waitForTask(swapTask.taskUid);

  // 7. Delete old temp index (now contains old data)
  await client.deleteIndex(tempIndexName);
  log(`Deleted index: ${tempIndexName}`);

  log(`\nReindex completed successfully for: ${indexName}`);
}

async function main() {
  log("=".repeat(60));
  log("Starting Aidbox Examples Indexer");
  log("=".repeat(60));

  // Step 1: Validate environment
  validateEnv();

  const meilisearchUrl = process.env.MEILISEARCH_HOST_URL!.replace(/\/$/, "");
  const apiKey = process.env.MEILISEARCH_API_KEY!;
  const githubToken = process.env.GITHUB_TOKEN!;

  // Initialize Meilisearch client
  const client = new MeiliSearch({
    host: meilisearchUrl,
    apiKey,
  });

  // Step 2: Fetch latest artifact
  const artifact = await fetchLatestArtifact(githubToken);
  if (!artifact) {
    log("No artifact to process, exiting");
    process.exit(0);
  }

  // Step 3: Download artifact ZIP
  const zipBytes = await downloadArtifact(githubToken, artifact.id);

  // Step 4: Extract metadata
  const examplesData = await extractMetadata(zipBytes);

  // Step 5: Transform data
  const documents = transformForMeilisearch(examplesData);

  // Step 6: Reindex with temp/swap pattern
  await reindexExamples(client, "examples", documents);

  log("=".repeat(60));
  log("Indexing completed successfully");
  log("=".repeat(60));
}

main().catch((error) => {
  log(`FATAL ERROR: ${error}`);
  process.exit(1);
});
